integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/management/CarbonInsertFromStageCommand.scala

```diff
@@ -0,0 +1,376 @@
+import java.util
+import java.util.Collections
+import java.util.concurrent.{Executors, ExecutorService}
+
+import scala.collection.JavaConverters._
+
+import com.google.gson.Gson
+import org.apache.hadoop.conf.Configuration
+import org.apache.log4j.Logger
+import org.apache.spark.sql.{CarbonEnv, Row, SparkSession}
+import org.apache.spark.sql.execution.command.{Checker, DataCommand}
+import org.apache.spark.sql.util.SparkSQLUtil
+
+import org.apache.carbondata.common.exceptions.sql.MalformedCarbonCommandException
+import org.apache.carbondata.common.logging.LogServiceFactory
+import org.apache.carbondata.core.constants.CarbonCommonConstants
+import org.apache.carbondata.core.datastore.filesystem.CarbonFile
+import org.apache.carbondata.core.datastore.impl.FileFactory
+import org.apache.carbondata.core.locks.{CarbonLockFactory, CarbonLockUtil, ICarbonLock, LockUsage}
+import org.apache.carbondata.core.metadata.SegmentFileStore
+import org.apache.carbondata.core.metadata.schema.table.CarbonTable
+import org.apache.carbondata.core.statusmanager.{SegmentStatus, SegmentStatusManager, StageInput}
+import org.apache.carbondata.core.util.path.CarbonTablePath
+import org.apache.carbondata.processing.loading.FailureCauses
+import org.apache.carbondata.processing.loading.model.CarbonLoadModel
+import org.apache.carbondata.processing.util.CarbonLoaderUtil
+import org.apache.carbondata.spark.load.DataLoadProcessBuilderOnSpark
+
+/**
+ * Collect stage input files and trigger a loading into carbon table.
+ *
+ * @param databaseNameOp database name
+ * @param tableName table name
+ */
+case class CarbonInsertFromStageCommand(
+    databaseNameOp: Option[String],
+    tableName: String
+) extends DataCommand {
+
+  @transient var LOGGER: Logger = _
+
+  override def processData(spark: SparkSession): Seq[Row] = {
+    LOGGER = LogServiceFactory.getLogService(this.getClass.getCanonicalName)
+    Checker.validateTableExists(databaseNameOp, tableName, spark)
+    val table = CarbonEnv.getCarbonTable(databaseNameOp, tableName)(spark)
+    val hadoopConf = spark.sessionState.newHadoopConf()
+    FileFactory.getConfiguration.addResource(hadoopConf)
+    setAuditTable(table)
+
+    if (!table.getTableInfo.isTransactionalTable) {
+      throw new MalformedCarbonCommandException("Unsupported operation on non transactional table")
+    }
+
+    val tablePath = table.getTablePath
+    val stagePath = CarbonTablePath.getStageDir(tablePath)
+    val snapshotFilePath = CarbonTablePath.getStageSnapshotFile(tablePath)
+    var loadModel: CarbonLoadModel = null
+    val lock = acquireIngestLock(table)
+
+    try {
+      // 1. Check whether we need to recover from previous failure
+      // We use a snapshot file to indicate whether there was failure in previous
+      // ingest operation. A Snapshot file will be created when an ingest operation
+      // starts and will be deleted only after the whole ingest operation is finished,
+      // which includes two actions:
```

Comment:
what happens if this command executed concurrently ? one instance thinks other is in-progress segment as recovery scenario. So, 
Need to block concurrent "insert into stage", just like how we have blocked "insert into overwrite"
can be handled in other open PR also, #3519

Reply:
I saw "acquireIngestLock" this now. It is already handled.